xarray/ufuncs.py

```diff
@@ -0,0 +1,125 @@
+"""xarray specific universal functions."""
+
+import textwrap
+import warnings
+
+import numpy as np
+
+import xarray as xr
+from xarray.core.groupby import GroupBy
+from xarray.namedarray.pycompat import array_type
+
+
+def _walk_array_namespaces(obj, namespaces):
+    if isinstance(obj, xr.DataTree):
+        for node in obj.subtree:
+            _walk_array_namespaces(node.dataset, namespaces)
+    elif isinstance(obj, xr.Dataset):
+        for name in obj.data_vars:
+            _walk_array_namespaces(obj[name], namespaces)
+    elif isinstance(obj, GroupBy):
+        _walk_array_namespaces(next(iter(obj))[1], namespaces)
+    elif isinstance(obj, xr.DataArray | xr.Variable):
+        _walk_array_namespaces(obj.data, namespaces)
+    elif isinstance(obj, array_type("dask")):
+        _walk_array_namespaces(obj._meta, namespaces)
```

Comment:
Why not sure Dask's `__array_namespace__` instead? That feels a little cleaner than special case logic for dask.array.